windows-agent/internal/config/config.go

```diff
@@ -156,27 +158,86 @@ func (c *Config) SetUserLandscapeConfig(ctx context.Context, landscapeConfig str
 		return errors.New("attempted to set a user-provided landscape configuration when there already is a higher priority one")
 	}
 
+	landscapeConfig, err := completeLandscapeConfig(landscapeConfig, c.Landscape.UID)
+	if err != nil {
+		return fmt.Errorf("config: could not complete Landscape configuration: %v", err)
+	}
+
 	isNew, err := c.set(&c.Landscape.UserConfig, landscapeConfig)
 	if err != nil {
 		return errors.New("config: could not set Landscape configuration")
 	}
 
 	if isNew {
-		c.notifyLandsape(ctx, landscapeConfig, c.Landscape.UID)
+		c.notifyLandscape(ctx, landscapeConfig, c.Landscape.UID)
 	}
 
 	return nil
 }
 
-// SetLandscapeAgentUID overrides the Landscape agent UID.
-func (c *Config) SetLandscapeAgentUID(uid string) error {
-	if _, err := c.set(&c.Landscape.UID, uid); err != nil {
-		return fmt.Errorf("config: could not set Landscape agent UID: %v", err)
+// SetLandscapeAgentUID overrides the Landscape agent UID and notify listeners.
+func (c *Config) SetLandscapeAgentUID(ctx context.Context, uid string) error {
+	isUpdated, conf, err := c.setAgentUIDAndUpdateClientConf(ctx, uid)
+	if err != nil {
+		return err
+	}
+
+	if isUpdated {
+		log.Debugf(ctx, "config: notifying Landscape config listeners about agent UID change: %s", uid)
+		c.notifyLandscape(ctx, conf, uid)
 	}
 
 	return nil
 }
 
+// setAgentUIDAndUpdateClientConf sets the agent UID and updates the client configuration, returning the new configuration and the 'true' boolean flag to indicate that the configuration has changed,
+// allowing the caller to trigger notifications, preventing that from happening while we hold a lock.
+func (c *Config) setAgentUIDAndUpdateClientConf(ctx context.Context, uid string) (bool, string, error) {
```

Comment:
In looking for flexibility I considered the fact that we could be transitioning into an empty Landscape client config. But that really makes no sense, because if we have a hostagent UID we have some Landscape config, so yeah, no need for that boolean.